### PR TITLE
feat(config): add configuration coverage validation script (fixes #4185)

### LINKED ISSUE #4185: [Config P2.1] Create configuration validation script
## Background

Need automated validation to ensure prod.md values match env.schema.yaml definitions.

## Blueprint Alignment

Section 4.6 Evidence Ledger - Configuration Audit Trail

## Tasks

- [ ] Create `scripts/validate_config_coverage.py`:
  - Parse env.schema.yaml for all defined variables
  - Compare against prod.md values
  - Report missing definitions
  - Report type mismatches
  - Report value discrepancies

- [ ] Add to CI workflow

## Acceptance Criteria

- [ ] Script validates all 77 prod.md values
- [ ] CI fails if configuration drift detected
- [ ] Clear error messages for each issue type

### PR DESCRIPTION
**Pull request by RC918**
## Summary

Adds a configuration validation script that compares production configuration values against the environment variable schema to detect configuration drift.

**New files:**
- `scripts/validate_config_coverage.py` - Validates prod.md against env.schema.yaml, reporting:
  - Variables in prod.md but not defined in schema (UNDEFINED)
  - Required variables missing from prod.md (MISSING_REQUIRED)
  - Type mismatches and invalid choices
  - Custom values that differ from defaults (INFO)
- `.github/workflows/config-validation.yml` - CI workflow that validates schema structure

**Blueprint Alignment:** Section 4.6 Evidence Ledger - Configuration Audit Trail

## Review & Testing Checklist for Human

- [ ] **Default file path**: The script defaults to `attachments/prod.md` which doesn't exist in the repo. Verify this is acceptable (script requires `--prod-file` argument for actual use)
- [ ] **CI workflow scope**: The workflow only validates schema structure, not full prod.md coverage (since prod.md contains sensitive values not in repo). Confirm this is the intended behavior
- [ ] **Type validator edge cases**: Review the integer validator logic for edge cases like empty strings or malformed values

**Test plan:**
```bash
# Run locally with actual prod.md file
python scripts/validate_config_coverage.py --prod-file /path/to/prod.md --schema-file config/env.schema.yaml

# Expected: Report showing 40 UNDEFINED variables (schema needs updating separately)
```

### Notes

Initial test run identified 40 variables in prod.md that are not defined in env.schema.yaml - this is real configuration drift that should be addressed in a follow-up issue to update the schema.

---
Link to Devin run: https://app.devin.ai/sessions/fb70960e8d714ce396d9ece377851f8d
Requested by: Ryan Chen (@RC918)

---

> [!NOTE]
> Adds automated configuration validation and basic CI checks to detect drift between production values and the env schema.
> 
> - New `scripts/validate_config_coverage.py` parses `attachments/prod.md` and `config/env.schema.yaml`, reporting `UNDEFINED`, `MISSING_REQUIRED`/`MISSING_SECRET`, type/choice mismatches, and `CUSTOM_VALUE` infos; includes CLI flags (`--prod-file`, `--schema-file`, `--strict`) and structured exit codes
> - New `.github/workflows/config-validation.yml` runs on changes to the schema/script to validate schema structure (presence of `version` and `fields`) and publishes a brief summary; instructs how to run full validation locally
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit a63a5630825b835a9fd663b0b3fd496e1f52b62d. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: scripts/validate_config_coverage.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+
+# Type mapping for validation
+TYPE_VALIDATORS = {
+    "string": lambda v: isinstance(v, str),
+    "integer": lambda v: v.isdigit() or (v.startswith("-") and v[1:].isdigit()),
+    "boolean": lambda v: v.upper() in ("TRUE", "FALSE", "YES", "NO", "1", "0"),
+    "url": lambda v: v.startswith(("http://", "https://", "redis://", "rediss://", "postgresql://")),
+    "secret": lambda v: isinstance(v, str),  # Secrets are just strings
+    "float": lambda v: _is_float(v),
+}
```
Comment (RC918) [inline scripts/validate_config_coverage.py]: IMPACT: The integer validation regex only accepts string representations of integers without leading zeros or formatting (e.g., '0123', '1,000'). This could cause false positives for valid numeric formats used in configs. The float validator delegates to _is_float which handles scientific notation and edge cases correctly, but integer parsing is more restrictive than necessary.

File: .github/workflows/config-validation.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+          with open('config/env.schema.yaml', 'r') as f:
+              schema = yaml.safe_load(f)
+          assert 'fields' in schema, 'Schema must have fields key'
+          assert 'version' in schema, 'Schema must have version key'
+          print(f'Schema version: {schema[\"version\"]}')
+          print(f'Total fields defined: {len(schema[\"fields\"])}')
+          required = sum(1 for f in schema['fields'].values() if f.get('required', False))
+          print(f'Required fields: {required}')
+          print('Schema structure validation: PASSED')
+          "
```
Comment (RC918) [inline .github/workflows/config-validation.yml]: IMPACT: The CI workflow only validates schema structure (presence of 'fields' and 'version') but does not validate individual field definitions. This could allow malformed field schemas (e.g., invalid types, missing required properties) to pass validation. The script has comprehensive validation but it's not being utilized in CI.

**Review by RC918 — COMMENTED** [review]
## MorningAI Code Review

File: scripts/validate_config_coverage.py (hunk 1 above)
Comment (RC918) [reply scripts/validate_config_coverage.py]: Valid observation. The integer validator is designed for typical env var values which are simple integers without formatting. However, I acknowledge edge cases exist:

- Leading zeros (e.g., `0123`) - could be octal or just padded
- Formatted numbers (e.g., `1,000`) - unlikely in env vars but possible

For this PR, the current validation is sufficient for the expected use case (env var values are typically simple integers). I'll create a follow-up issue to improve type validation robustness if needed.

The float validator already handles edge cases correctly via `float()` parsing.

File: scripts/validate_config_coverage.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        action="store_true",
+        help="Treat warnings as errors",
+    )
+
+    args = parser.parse_args()
+
+    try:
+        # Parse files
+        prod_config = parse_prod_md(args.prod_file)
+        schema = parse_schema(args.schema_file)
```
[the remark this answers is not in the document]
Comment (RC918) [reply scripts/validate_config_coverage.py]: Created follow-up issue #4242 for path traversal protection. This is a P3 low-priority item since the script is a CLI tool typically run locally, but good security practice for automated environments.

File: scripts/validate_config_coverage.py [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+        var_name: Variable name (for error messages)
+        value: The value to validate
+        choices: List of allowed values
+
+    Returns:
+        Tuple of (is_valid, error_message)
+    """
+    # Normalize for case-insensitive comparison
+    normalized_value = value.lower()
+    normalized_choices = [c.lower() for c in choices]
```
[the remark this answers is not in the document]
Comment (RC918) [reply scripts/validate_config_coverage.py]: Addressed in commit f217c937: Now using `str(c).lower()` to convert choices to strings before comparison, handling YAML-parsed bool/int types.

File: scripts/validate_config_coverage.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+    prod_vars = set(prod_config.keys())
+
+    # 1. Check for variables in prod.md but not in schema
+    undefined_vars = prod_vars - schema_vars
+    for var in sorted(undefined_vars):
+        errors.append(f"UNDEFINED: {var} is in prod.md but not defined in schema")
+
+    # 2. Check for required variables missing from prod.md
+    for var_name, var_schema in schema.items():
+        if var_schema.get("required", False) and var_name not in prod_vars:
```
[the remark this answers is not in the document]
Comment (RC918) [reply scripts/validate_config_coverage.py]: Addressed in commit f217c937: Added null checks for individual field definitions. If a field has a null definition (e.g., `SOME_VAR:` with no nested content), the script now:
1. Logs a `MALFORMED_SCHEMA` warning
2. Skips the field gracefully instead of crashing
